crates/bevy_math/src/bezier.rs

```diff
@@ -0,0 +1,376 @@
+    /// ```
+    ///
+    /// To accomplish this, we need to be able to find the position `y` on a Bezier curve, given the
+    /// `x` value. As discussed in the [`Bezier`] documentation, a Bezier curve is an implicit
+    /// parametric function like B(t) = (x,y). To find `y`, we first solve for `t` that corresponds
+    /// to the given `x` (`time`). We use the Newton-Raphson root-finding method to quickly find a
+    /// value of `t` that matches `x`. Once we have this we can easily plug that `t` into our
+    /// Bezier's `position` function, to find the `y` component, which is how far along our
+    /// animation should be. In other words:
+    ///
+    /// > Given `time` in `0..=1`
+    ///
+    /// > Use Newton's method to find a value of `t` that results in B(t) = (x,y) where `x == time`
+    ///
+    /// > Once a solution is found, use the resulting `y` value as the final result
+    ///
+    /// # Performance
+    ///
+    /// This operation can be used frequently without fear of performance issues. Benchmarks show
+    /// this operation taking on the order of 50 nanoseconds.
+    pub fn ease(&self, time: f32) -> f32 {
+        let x = time.clamp(0.0, 1.0);
+        let t = self.find_t_given_x(x);
+        self.evaluate_y_at(t)
+    }
+
+    /// Compute the x-coordinate of the point along the Bezier curve at `t`.
+    #[inline]
+    fn evaluate_x_at(&self, t: f32) -> f32 {
+        generic::position([0.0, self.p1.x, self.p2.x, 1.0], t)
+    }
+
+    /// Compute the y-coordinate of the point along the Bezier curve at `t`.
+    #[inline]
+    fn evaluate_y_at(&self, t: f32) -> f32 {
+        generic::position([0.0, self.p1.y, self.p2.y, 1.0], t)
+    }
+
+    /// Compute the slope of the line at the given parametric value `t` with respect to t.
+    #[inline]
+    fn dx_dt(&self, t: f32) -> f32 {
+        generic::velocity([0.0, self.p1.x, self.p2.x, 1.0], t)
+    }
+
+    /// Solve for the parametric value `t` that corresponds to the given value of `x` using the
+    /// Newton-Raphson method. See documentation on [`Self::ease`] for more details.
+    #[inline]
+    fn find_t_given_x(&self, x: f32) -> f32 {
+        // PERFORMANCE NOTE:
+        //
+        // I tried pre-solving and caching 11 values along the curve at struct instantiation in an
+        // attempt to give the solver a better starting guess. This ended up being slightly slower,
+        // possibly due to the increased size of the type. Another option would be to store the last
+        // `t`, and use that, however it's possible this could end up in a bad state where t is very
+        // far from the naive but generally safe guess of x, e.g. after an animation resets.
+        //
+        // Further optimization might not be needed however - benchmarks are showing it takes about
+        // 50 nanoseconds for an ease operation on my modern laptop, which seems sufficiently fast.
+        let mut t_guess = x;
+        (0..Self::MAX_ITERS).any(|_| {
```

Comment:
This is a pretty surprising way of doing control flow. I feel it would be a lot easier to understand if this was rewritten to use `for _ in 0..SELF::MAX_ITERS`, and then using `break;` instead of returning `true`

Reply:
The reason for this was to improve the chances of loop unrolling. Iterators are more likely to be properly unrolled, and the limited benchmarking I did suggested the iterator version was slightly faster. I spent quite a while attempting to optimize this function.

Reply:
I revisted benchmarks, and didn't find a difference, so I reverted back to a simpler for loop